### DISCUSSION TITLE
AMReX Hypre supported boundary conditions

### DISCUSSION THREAD

**Discussion by ankithadas** [AMReX-Codes/amrex#3508] [Q&A] [answered]
Hey AMReX team,

I just wanted to confirm my finding after going through the HYPRE wrapper and make sure I haven't missed anything. To my understanding, does the Hypre Wrapper only support zero Dirichlet and zero gradient Nuemann boundary conditions.  

Kind regards,
Ankith

**Comment by WeiqunZhang** [comment, marked as answer]
It's true that the Hypre wrapper at Src/Extern/HYPRE only supports homogeneous Dirichlet and Neumann BCs. However, this does not mean it cannot be used to solve problems with inhomogeneous Dirichlet and Neumann BCs. The inhomogeneous part is handled in src/LinearSolvers/MLMG.

**Reply by ankithadas** [reply]
Could you please direct me where the inhomogeneous Dirichlet and Neumann BC components are added to the RHS in MLMG solver? I am attempting to write a linear solver that can handle advection as well

**Reply by WeiqunZhang** [reply]
https://github.com/AMReX-Codes/amrex/blob/d4d0e9031d036070fad6d1d3bc96a8a32b48b3a9/Src/LinearSolvers/MLMG/AMReX_MLCellABecLap.H#L275 Here is an example of modifying RHS for inhomogeneous Neumann.

https://github.com/AMReX-Codes/amrex/blob/d4d0e9031d036070fad6d1d3bc96a8a32b48b3a9/Src/LinearSolvers/MLMG/AMReX_MLABecLaplacian.H#L451 Here is how we handle Robin BC.

As for inhomogeneous Dirichlet, it does not change RHS. It's handled here. https://github.com/AMReX-Codes/amrex/blob/d4d0e9031d036070fad6d1d3bc96a8a32b48b3a9/Src/LinearSolvers/MLMG/AMReX_MLCellLinOp.H#L652 Note that given a linear system `L(phi) = f`, we compute `res = f - L(phi_initial_guess)`. Then we solve `L(dphi) = res`, for which Dirichlet BC becomes homogenous. When we use Hypre, what it solves is actually `L(dphi) = res` and the final solution is then `phi_initial_guess + dphi`.

**Reply by WeiqunZhang** [reply]
Also note that there are a number of classes in Src/Extern/HYPRE. For the HypreSolver class, there is no concept of BC. It's algebraic. It's up to the user to set up the matrix.
